modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/validation/MediaTypeModelValidatorTestCase.java

```diff
@@ -0,0 +1,312 @@
+    when(sourceModel.getModelProperty(ExtensionTypeDescriptorModelProperty.class)).thenReturn(of(descriptorModelProperty));
+    when(operationModel.getModelProperty(ExtensionOperationDescriptorModelProperty.class))
+        .thenReturn(of(operationDescriptorModelProperty));
+
+    when(sourceModel.getModelProperty(MediaTypeModelProperty.class)).thenReturn(empty());
+    when(operationModel.getModelProperty(MediaTypeModelProperty.class)).thenReturn(empty());
+
+    when(operationModel.getOutput()).thenReturn(operationOutputModel);
+    when(sourceModel.getOutput()).thenReturn(sourceOutputModel);
+
+    when(operationOutputModel.getType()).thenReturn(toMetadataType(Object.class));
+    when(sourceOutputModel.getType()).thenReturn(toMetadataType(Object.class));
+
+    when(sourceModel.getName()).thenReturn("source");
+    when(operationModel.getName()).thenReturn("operation");
+
+    mockExtensionOperationDescriptorModelProperty("returnsObject");
+    mockExtensionTypeDescriptorModelProperty(TestObjectSource.class);
+
+    setCompileTime(true);
+    visitableMock(operationModel, sourceModel);
+  }
+
+  private void setCompileTime(boolean compileTime) {
+    when(extensionModel.getModelProperty(CompileTimeModelProperty.class))
+        .thenReturn(ofNullable(compileTime ? new CompileTimeModelProperty() : null));
+  }
+
+  private void mockMediaTypeAnnotation(ConnectableComponentModel model, String value, boolean strict) {
+    when(model.getModelProperty(MediaTypeModelProperty.class)).thenReturn(ofNullable(new MediaTypeModelProperty(value, strict)));
+  }
+
+  private void mockExtensionTypeDescriptorModelProperty(Class sourceClass) {
+    SourceTypeWrapper sourceType = new SourceTypeWrapper(sourceClass, typeLoader);
+    ExtensionTypeDescriptorModelProperty extensionTypeDescriptorModelProperty =
+        mock(ExtensionTypeDescriptorModelProperty.class, new ReturnsDeepStubs());
+    when(sourceModel.getModelProperty(ExtensionTypeDescriptorModelProperty.class))
+        .thenReturn(of(extensionTypeDescriptorModelProperty));
+    when(extensionTypeDescriptorModelProperty.getType())
+        .thenReturn(sourceType);
+  }
+
+  private void mockExtensionOperationDescriptorModelProperty(String operationMethodName) throws NoSuchMethodException {
+    Method operationMethod = TestMethods.class.getMethod(operationMethodName);
+    Type operationReturnType = new TypeWrapper(ResolvableType.forMethodReturnType(operationMethod), typeLoader);
+    ExtensionOperationDescriptorModelProperty extensionOperationDescriptorModelProperty =
+        mock(ExtensionOperationDescriptorModelProperty.class, new ReturnsDeepStubs());
+    when(operationModel.getModelProperty(ExtensionOperationDescriptorModelProperty.class))
+        .thenReturn(of(extensionOperationDescriptorModelProperty));
+    when(extensionOperationDescriptorModelProperty.getOperationReturnType()).thenReturn(operationReturnType);
+  }
+
+  @Test(expected = IllegalModelDefinitionException.class)
```

Comment:
I know it's a pain, but use ExpectedException Rule and match by exception type _and_ error message substring. Otherwise it may fail for a different reason than the expected for this test. Take a look at `org.mule.runtime.module.extension.internal.loader.validation.MetadataComponentModelValidatorTestCase`